lib/cereal/cereal.go

```diff
@@ -497,19 +503,188 @@ type TaskExecutorOpts struct {
 	Workers int
 }
 
+// registeredExecutor is responsible for polling for and executing
+// tasks using the given task executor.
+//
+// Tasks are polled for based on
+//
+// - a time-based interval, and
+// - in-process notifications that are sent when we think there
+// might be a task available.
+//
+// When either of these tiggers occur, we spawn a new goroutine (the
+// "task worker") provided that we are under the maximum configured
+// workers for this task executor.
+//
+// The task worker will repeatedly dequeue tasks until none are
+// available. To ensure our queue grows in response to load, if a task
+// is found in the queue, another task worker will be spawned.
+//
 type registeredExecutor struct {
+	name     string
 	executor TaskExecutor
 	opts     TaskExecutorOpts
+
+	wakeupChan chan struct{}
+
+	maxWorkers    int
+	activeWorkers int
+	sync.Mutex    // guards activeWorkers
+
+	wg sync.WaitGroup
+}
+
+// DecActiveWorkers decrements the worker count, clamping it at zero.
+func (r *registeredExecutor) DecActiveWorkers() {
+	r.Lock()
+	defer r.Unlock()
+
+	if r.activeWorkers > 0 {
+		r.activeWorkers--
+	}
+}
+
+// IncActiveWorkers increments the active worker count if another
+// worker is allowed. It returns true if the count was incremented and
+// false otherwise. The caller should not start a worker if false is
+// returned.
+func (r *registeredExecutor) IncActiveWorkers() bool {
+	r.Lock()
+	defer r.Unlock()
+
+	if r.activeWorkers < r.maxWorkers {
+		r.activeWorkers++
+		return true
+	}
+
+	return false
+}
+
+// WakeupPoller will wake up the poller for this
+// registeredExecutor. This is called when a workflow executor
+// enqueues tasks.
+func (r *registeredExecutor) WakeupPoller() {
+	select {
+	case r.wakeupChan <- struct{}{}:
+	default:
+	}
+}
+
+// StartPoller runs until the given context is cancelled, starting a
+// TaskWorker every interval (or sooner if we've been notified of a
+// change).
+func (r *registeredExecutor) StartPoller(ctx context.Context, b backend.Driver, taskPollInterval time.Duration, workflowWakeupFun func()) {
+	logctx := logrus.WithField("task_name", r.name)
+	logctx.Infof("Starting task poller")
+	for {
+		select {
+		case <-ctx.Done():
+			logctx.WithField("active_workers", r.activeWorkers).Info("Waiting for all task workers to exit")
+			r.wg.Wait()
+			logctx.Info("Exiting task poller")
+			return
+		case <-time.After(taskPollInterval):
```

Comment:
i think this leak might matter depending on how much wakeupChan gets a message

Reply:
Correct, to avoid the leak we'll need to use a `time.Timer`, which is a bit of a pain to use correctly, but I can wire it up.